### PR TITLE
Adding in disableGlobalProjectWarning, fixes #1168

### LINKED ISSUE #1168: Remove "unable to detect a global setting for GCP Project" warning.
https://github.com/pulumi/pulumi-gcp/blob/f0115d920cb100c5fc4f39abfb54d90bf6146b61/provider/resources.go#L318

I have pulumi configuration which creates GCP projects (thus no global project ID is set) and I've found this warning very annoying in my outputs.

Could we remove or supress this specific warning?

### PR DESCRIPTION
**Pull request by rshade**
Adds `disableGlobalProjectWarning` to the ExtraConfig for the provider, allowing users to disable the initial default project. This fixes #1168

### CODE REVIEW

File: provider/errors/no_project.txt [hunk 1]
```diff
@@ -1,4 +1,6 @@
 unable to detect a global setting for GCP Project.
 Pulumi will rely on per-resource settings for this operation.
 Set the GCP Project by using:
-    `pulumi config set gcp:project <project>`
\ No newline at end of file
+    `pulumi config set gcp:project <project>`
+If you would like to override global project setting:
```
Comment (VenelinMartinov) [inline provider/errors/no_project.txt:5]:
```suggestion
If you would like to disable this warning use:
```

File: provider/provider_yaml_test.go [hunk 2]
```diff
@@ -80,12 +80,11 @@ func TestComputeFirewallUpgrade(t *testing.T) {
 }
 
 func TestCloudFunctionUpgrade(t *testing.T) {
```
Comment (VenelinMartinov) [inline provider/provider_yaml_test.go:82]: The issue is not closed why are we removing the skip here?

File: provider/provider_yaml_test.go [hunk 3]
```diff
@@ -80,12 +80,11 @@ func TestComputeFirewallUpgrade(t *testing.T) {
...
-	t.Skipf("TODO[https://github.com/pulumi/pulumi-gcp/issues/2670]: Flaky")
 	testUpgrade(t, "test-programs/cloudfunctions-function")
 }
 
 func TestNetworkUpgrade(t *testing.T) {
-	t.Skipf("TODO[https://github.com/pulumi/pulumi-gcp/issues/1655]: Flaky")
+	t.Skipf("Flakey: see https://github.com/pulumi/pulumi-gcp/issues/1655 for details")
```
Comment (VenelinMartinov) [inline provider/provider_yaml_test.go:87]: Why are we removing the TODO?

File: provider/provider_yaml_test.go [hunk 4]
```diff
@@ -115,7 +114,6 @@ func TestIamMember(t *testing.T) {
 }
 
 func TestLogSink(t *testing.T) {
-	t.Skipf("TODO[https://github.com/pulumi/pulumi-gcp/issues/2671]: Flaky")
```
Comment (VenelinMartinov) [inline provider/provider_yaml_test.go:118]: This is unrelated and does not look correct, as the issues are still open. Can you please remove these changes?

File: provider/provider_yaml_test.go [hunk 5]
```diff
@@ -211,6 +209,21 @@ func TestNoGlobalProjectWarning(t *testing.T) {
...
+	}
+	cwd, err := os.Getwd()
+	require.NoError(t, err)
+	proj := getProject()
+	test := pulumitest.NewPulumiTest(t, "test-programs/project-bucket",
+		opttest.LocalProviderPath(providerName, filepath.Join(cwd, "..", "bin")))
+
+	test.SetConfig(t, "gcp:disableGlobalProjectWarning", "true")
+	test.SetConfig(t, "gcpProj", proj)
+	test.Up(t)
```
Comment (VenelinMartinov) [inline provider/provider_yaml_test.go:224]: You need to capture the stdout from the result of `test.Up(t)` here and assert that it does not contain the warning.

File: provider/resources.go [hunk 6]
```diff
@@ -432,17 +427,23 @@ func preConfigureCallbackWithLogger(credentialsValidationRun *atomic.Bool, gcpCl
...
+
+		if !disableGlobalProjectWarning {
+			// validate the gcloud config
+			err := config.LoadAndValidate(ctx)
+			if err != nil {
+				return fmt.Errorf(noCredentialsErr, err)
+			}
+		}
+
+		if !skipRegionValidation && disableGlobalProjectWarning && config.Region != "" && config.Project != "" {
```
Comment (VenelinMartinov) [inline provider/resources.go:446]: This looks wrong. Did you mean `! disableGlobalProjectWarning`.

I think an early return on line 437 would be more readable:

```golang
if disableGlobalProjectWarning {
    return
}
```

**Comment by VenelinMartinov** [comment]
Thanks for writing this up! I had a few questions here but we're almost there.

File: provider/provider_yaml_test.go [hunk 7]
```diff
@@ -211,6 +211,22 @@ func TestNoGlobalProjectWarning(t *testing.T) {
...
+		t.Skipf("Skipping in testing.Short() mode, assuming this is a CI run without GCP creds")
+	}
+	cwd, err := os.Getwd()
+	require.NoError(t, err)
+	proj := getProject()
+	test := pulumitest.NewPulumiTest(t, "test-programs/project-bucket",
+		opttest.LocalProviderPath(providerName, filepath.Join(cwd, "..", "bin")))
+
+	test.SetConfig(t, "gcp:disableGlobalProjectWarning", "true")
+	test.SetConfig(t, "gcpProj", proj)
```
[the remark this answers is not in the document]
Comment (rshade) [reply provider/provider_yaml_test.go:225]: That is a different config key right, which is necessary for the defined non-default provider.

[the remark this answers is not in the document]
Comment (VenelinMartinov) [reply provider/provider_yaml_test.go:225]: True, good point, thank you

File: provider/resources.go [hunk 8]
```diff
@@ -432,17 +427,23 @@ func preConfigureCallbackWithLogger(credentialsValidationRun *atomic.Bool, gcpCl
...
-
 		skipRegionValidation := tfbridge.ConfigBoolValue(
 			vars, "skipRegionValidation", []string{"PULUMI_GCP_SKIP_REGION_VALIDATION"},
 		)
 
-		if !skipRegionValidation && config.Region != "" && config.Project != "" {
+		disableGlobalProjectWarning := tfbridge.ConfigBoolValue(
+			vars, "disableGlobalProjectWarning", []string{"PULUMI_GCP_DISABLE_GLOBAL_PROJECT_WARNING"},
+		)
+
```
Comment (VenelinMartinov) [inline provider/resources.go:437]: Can we add an early return here?

```golang
if disableGlobalProjectWarning {
    return
}
```

File: provider/resources.go [hunk 9]
```diff
@@ -432,17 +427,25 @@ func preConfigureCallbackWithLogger(credentialsValidationRun *atomic.Bool, gcpCl
...
 		if err != nil {
 			return fmt.Errorf(noCredentialsErr, err)
 		}
 
-		skipRegionValidation := tfbridge.ConfigBoolValue(
-			vars, "skipRegionValidation", []string{"PULUMI_GCP_SKIP_REGION_VALIDATION"},
-		)
-
-		if !skipRegionValidation && config.Region != "" && config.Project != "" {
+		if !skipRegionValidation && disableGlobalProjectWarning && config.Region != "" && config.Project != "" {
```
Comment (VenelinMartinov) [inline provider/resources.go:448]: The change here does not look necessary any more, can we remove that?

**Review by VenelinMartinov — APPROVED** [review]
Great, thank you! 💯

**Review by iwahbe — APPROVED** [review]
LGTM

If you want to go above and beyond, you could document this in https://github.com/pulumi/pulumi-gcp/blob/master/docs/installation-configuration.md.

**Comment by VenelinMartinov** [comment]
@iwahbe this might be hitting a bridge bug - the extra config seems to be incorrectly passed to the TF provider. I'm investigating

**Comment by VenelinMartinov** [comment]
Opened https://github.com/pulumi/pulumi-terraform-bridge/issues/2702 for the issue and I also have a fix in progress.

**Comment by rshade** [comment]
@VenelinMartinov @iwahbe can one of yall looking under the doc and see if that looks good. If so its ready to merge.

**Review by VenelinMartinov — APPROVED** [review]
Great change, thank you!
